File: app/src/main/java/pl/sokols/shoppinglist/utils/ShopItemDialog.java
package pl.sokols.shoppinglist.utils;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;

import org.jetbrains.annotations.NotNull;

import java.util.Objects;

import pl.sokols.shoppinglist.R;
import pl.sokols.shoppinglist.data.entities.ShopItem;
import pl.sokols.shoppinglist.databinding.ShopItemDialogBinding;

public class ShopItemDialog extends DialogFragment {

    private final ShopItem shopItem;
    private final OnItemClickListener listener;
    private ShopItemDialogBinding dialogBinding;

    public ShopItemDialog(int shopListId, OnItemClickListener listener) {
        this.shopItem = new ShopItem("", 0, shopListId, false);
        this.listener = listener;
    }

    @Nullable
    @org.jetbrains.annotations.Nullable
    @Override
    public View onCreateView(@NonNull @NotNull LayoutInflater inflater, @Nullable @org.jetbrains.annotations.Nullable ViewGroup container, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        dialogBinding = ShopItemDialogBinding.inflate(inflater, container, false);
        dialogBinding.setShopItem(shopItem);
        setComponents();
        return dialogBinding.getRoot();
    }

    private void setComponents() {
        dialogBinding.applyDialogButton.setOnClickListener(v -> {
            System.out.println(shopItem);
            if (dialogBinding.itemNameEditText.getText().toString().isEmpty()) {
                dialogBinding.itemNameTextInputLayout.setError(getString(R.string.cannot_be_empty));
                dialogBinding.itemNameTextInputLayout.setErrorEnabled(true);
            } else {
                dialogBinding.itemNameTextInputLayout.setErrorEnabled(false);
            }
            if (dialogBinding.itemAmountEditText.getText().toString().isEmpty()) {
                dialogBinding.itemAmountTextInputLayout.setError(getString(R.string.cannot_be_empty));
                dialogBinding.itemAmountTextInputLayout.setErrorEnabled(true);
            } else {
                dialogBinding.itemAmountTextInputLayout.setErrorEnabled(false);
            }

            if (shopItem.getAmount() != null && shopItem.getName() != null
                    && shopItem.getAmount() != 0 && !shopItem.getName().isEmpty()) {
                listener.onItemClickListener(shopItem);
                Objects.requireNonNull(this.getDialog()).hide();
            }
        });
    }
}

File: app/src/main/java/pl/sokols/shoppinglist/utils/ShopItemDialog.kt
package pl.sokols.shoppinglist.utils

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import pl.sokols.shoppinglist.R
import pl.sokols.shoppinglist.data.entities.ShopItem
import pl.sokols.shoppinglist.databinding.ShopItemDialogBinding

/**
 * Helper class created to prepare dialog for ShopItems.
 */
class ShopItemDialog(
    sentShopItem: ShopItem?,
    shopListId: Int,
    private val listener: OnItemClickListener
) : DialogFragment() {

    var shopItem: ShopItem =
        sentShopItem ?: ShopItem(name = "", shopListId = shopListId, amount = 1)

    private lateinit var dialogBinding: ShopItemDialogBinding

    override fun onPause() {
        super.onPause()
        dismiss()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        dialogBinding = ShopItemDialogBinding.inflate(inflater, container, false)
        dialogBinding.shopItem = shopItem
        setComponents()
        return dialogBinding.root
    }

    /**
     * Method which:
     * - validates the inputs,
     * - displays the errors if the inputs are incorrect,
     * - turns on the listener if the inputs are correct.
     */
    private fun setComponents() {
        dialogBinding.applyDialogButton.setOnClickListener {
            if (shopItem.name.trim().isEmpty()) {
                dialogBinding.itemNameTextInputLayout.error = getString(R.string.incorrect_value)
                dialogBinding.itemNameTextInputLayout.isErrorEnabled = true
            } else {
                dialogBinding.itemNameTextInputLayout.isErrorEnabled = false
            }
            if (shopItem.amount == 0) {
                dialogBinding.itemAmountTextInputLayout.error = getString(R.string.incorrect_value)
                dialogBinding.itemAmountTextInputLayout.isErrorEnabled = true
            } else {
                dialogBinding.itemAmountTextInputLayout.isErrorEnabled = false
            }
            if (shopItem.name.trim().isNotEmpty() && shopItem.amount != 0) {
                listener.onItemClickListener(shopItem)
                dismiss()
            }
        }
    }
}
